Crud Operations Failing for an Map with enum as key

We have an embedded entity which uses MAP, The key for Map is  Java Enum. While reading(findById) and writing(save) this entity we are receiving a converison exception.

From Spring docs it suggest Enum is supported as a datatype , so can you help understand what's going wrong here.

https://cloud.spring.io/spring-cloud-static/spring-cloud-gcp/1.1.0.RC2/multi/multi__spring_data_cloud_datastore.html

#### Environment details
1. Java version: 8
2. spring-cloud-datastore: 1.1.0.RELEASE

#### Code example

  ``` 
 @Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Entity(name = "organization_preferences")
public class OrganizationPreferencesDto {
    @Id
    @Field(name = "organization_id")
    private String organizationId;

    @Field(name = "events")
    private Map<String, EventDto> eventNameToEventMapping;

}

``` 
  ```
public class EventDto {
    @Field(name = "event_name")
    private String eventName;

    @Field(name = "preferred_channels")
    private List<CommunicationChannels> preferredChannels;

    @Field(name = "preferences")
    private Map<CommunicationChannels, List<String>> channelNameToPreferences;

}
``` 
```
public enum CommunicationChannels {

	EMAIL(ChannelType.EMAIL, "EMAIL"), SMS(ChannelType.SMS, "SMS");
	
	private ChannelType grpcChannelFormat;
	private String channelName;

	private CommunicationChannels(final ChannelType grpcChannelFormat, final String channelName) {
		this.grpcChannelFormat = grpcChannelFormat;
		this.channelName = channelName;
	}

	public String getChannelName() {
		return channelName;
	}


	public ChannelType getGrpcChannelFormat() {
		return grpcChannelFormat;
	}
}

``` 




#### Stack trace
```
2019-05-22 03:01:10.365  WARN 12586 --- [          Gax-8] c.g.cloud.pubsub.v1.MessageDispatcher    : MessageReceiver failed to processes ack ID: K0MLKlgRTgQhIT4wPkVTRFAGFixdRkhRNxkIaFEOT14jPzUgKEURIysdAGNfNBsNaFxcdAEAWRp7LDRxY1kZAQhFe059ZB0Oa1pdcQZZBRx2e2R1aXLN1YuE6bfgagExULPX0dBkLcny58FEZh09WBJLLD5-LTlFQV5AEkwmDERJUw, the message will be nacked.

org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write organization_preferences.events; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write eventDto.preferences; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:149) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.lambda$doWithColumnBackedProperties$0(DatastorePersistentEntityImpl.java:110) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.data.mapping.model.BasicPersistentEntity.doWithProperties(BasicPersistentEntity.java:355) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.doWithColumnBackedProperties(DatastorePersistentEntityImpl.java:107) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:133) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:47) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate.convertToEntityForSave(DatastoreTemplate.java:310) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.DatastoreTemplate.save(DatastoreTemplate.java:115) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.repository.support.SimpleDatastoreRepository.save(SimpleDatastoreRepository.java:80) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.data.repository.core.support.RepositoryComposition$RepositoryFragments.invoke(RepositoryComposition.java:359) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryComposition.invoke(RepositoryComposition.java:200) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$ImplementationMethodExecutionInterceptor.invoke(RepositoryFactorySupport.java:644) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.doInvoke(RepositoryFactorySupport.java:608) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.lambda$invoke$3(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.data.repository.core.support.RepositoryFactorySupport$QueryExecutorMethodInterceptor.invoke(RepositoryFactorySupport.java:595) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.projection.DefaultMethodInvokingMethodInterceptor.invoke(DefaultMethodInvokingMethodInterceptor.java:59) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:93) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.data.repository.core.support.SurroundingTransactionDetectorMethodInterceptor.invoke(SurroundingTransactionDetectorMethodInterceptor.java:61) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at com.sun.proxy.$Proxy66.save(Unknown Source) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:343) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:198) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139) ~[spring-tx-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212) ~[spring-aop-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at com.sun.proxy.$Proxy66.save(Unknown Source) ~[na:na]
	at com.kramphub.postino.organization.preferences.OrganizationPreferencesService.insertByOrganizationId(OrganizationPreferencesService.java:65) ~[classes/:na]
	at com.kramphub.postino.organization.CreateOrganizationPubSubReader.processMessage(CreateOrganizationPubSubReader.java:51) ~[classes/:na]
	at com.kramphub.postino.organization.CreateOrganizationPubSubReader.lambda$0(CreateOrganizationPubSubReader.java:90) ~[classes/:na]
	at com.google.cloud.pubsub.v1.MessageDispatcher$4.run(MessageDispatcher.java:438) ~[google-cloud-pubsub-1.59.0.jar:1.59.0]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]
Caused by: org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to write eventDto.preferences; nested exception is org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:149) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.lambda$doWithColumnBackedProperties$0(DatastorePersistentEntityImpl.java:110) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.data.mapping.model.BasicPersistentEntity.doWithProperties(BasicPersistentEntity.java:355) ~[spring-data-commons-2.1.2.RELEASE.jar:2.1.2.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.mapping.DatastorePersistentEntityImpl.doWithColumnBackedProperties(DatastorePersistentEntityImpl.java:107) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:133) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.write(DefaultDatastoreEntityConverter.java:47) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbedded$5(TwoStepsConversions.java:303) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbedded(TwoStepsConversions.java:302) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$3(TwoStepsConversions.java:255) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbeddedMap$4(TwoStepsConversions.java:294) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbeddedMap(TwoStepsConversions.java:288) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$2(TwoStepsConversions.java:251) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:238) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:137) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	... 49 common frames omitted
Caused by: org.springframework.cloud.gcp.data.datastore.core.mapping.DatastoreDataException: Unable to convert class com.kramphub.postino.organization.preferences.CommunicationChannels to class java.lang.String
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnReadSingle(TwoStepsConversions.java:231) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWriteSingleEmbeddedMap$4(TwoStepsConversions.java:291) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.applyEntityValueBuilder(TwoStepsConversions.java:282) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWriteSingleEmbeddedMap(TwoStepsConversions.java:288) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.lambda$convertOnWrite$2(TwoStepsConversions.java:251) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:275) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.TwoStepsConversions.convertOnWrite(TwoStepsConversions.java:238) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	at org.springframework.cloud.gcp.data.datastore.core.convert.DefaultDatastoreEntityConverter.lambda$write$1(DefaultDatastoreEntityConverter.java:137) ~[spring-cloud-gcp-data-datastore-1.1.0.RELEASE.jar:1.1.0.RELEASE]
	... 66 common frames omitted


```

#### Additional information.	
We have also tried to use custom converters for same but it did not help while writing
```
public static final Converter<CommunicationChannels, String> COMMCHANNEL_STRING_CONVERTER = new Converter<CommunicationChannels, String>() {

		@Override
		public String convert(CommunicationChannels channel) {
			return channel.name();
		}
	};
	public static final Converter<String, CommunicationChannels> STRING_COMMCHANNEL_CONVERTER = new Converter<String, CommunicationChannels>() {

		@Override
		public CommunicationChannels convert(String channel) {
			if (channel.equalsIgnoreCase(CommunicationChannels.EMAIL.toString()))
				return CommunicationChannels.EMAIL;
			else
				return CommunicationChannels.SMS;
		}
	};

@Bean
	public DatastoreCustomConversions datastoreCustomConversions() {
		return new DatastoreCustomConversions(Arrays.asList(CommunicationChannels.COMMCHANNEL_STRING_CONVERTER,
				CommunicationChannels.STRING_COMMCHANNEL_CONVERTER));
	}
```


Thanks!

--- a/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java
+++ b/spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/convert/TwoStepsConversions.java
@@ -308,7 +308,7 @@ public class TwoStepsConversions implements ReadWriteConversions {
 		return applyEntityValueBuilder(null, kindName, (builder) -> {
 			Map map = (Map) val;
 			for (Object key : map.keySet()) {
-				String field = convertOnReadSingle(key,
+				String field = convertOnReadSingle(convertOnWriteSingle(key).get(),
 						ClassTypeInformation.from(String.class));
 				builder.set(field,
 						convertOnWrite(map.get(key),

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverterTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/DefaultDatastoreEntityConverterTests.java
@@ -653,10 +653,14 @@ public class DefaultDatastoreEntityConverterTests {
 		nestedInnerEmbeddedMap.put(1L, mapSimpleValues);
 		nestedEmbeddedMap.put("outer1", nestedInnerEmbeddedMap);
 
+		Map<TestDatastoreItem.Color, String> enumKeysMap = new HashMap<>();
+		enumKeysMap.put(TestDatastoreItem.Color.BLACK, "black");
+		enumKeysMap.put(TestDatastoreItem.Color.WHITE, "white");
+
 		TestItemWithEmbeddedEntity item = new TestItemWithEmbeddedEntity(123,
 				new EmbeddedEntity("abc"), embeddedEntities, mapSimpleValues,
 				mapListValues, embeddedEntityMapEmbeddedEntity,
-				embeddedEntityMapListOfEmbeddedEntities);
+				embeddedEntityMapListOfEmbeddedEntities, enumKeysMap);
 
 		item.setNestedEmbeddedMaps(nestedEmbeddedMap);
 

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/TestItemWithEmbeddedEntity.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/core/convert/TestItemWithEmbeddedEntity.java
@@ -46,12 +46,15 @@ public class TestItemWithEmbeddedEntity {
 
 	private Map<String, Map<Long, Map<String, String>>> nestedEmbeddedMaps;
 
+	Map<TestDatastoreItem.Color, String> mapEnumKeys;
+
 	public TestItemWithEmbeddedEntity(int intField, EmbeddedEntity embeddedEntityField,
 			List<EmbeddedEntity> listOfEmbeddedEntities,
 			Map<String, String> embeddedMapSimpleValues,
 			Map<String, String[]> embeddedMapListOfValues,
 			Map<String, EmbeddedEntity> embeddedEntityMapEmbeddedEntity,
-			Map<String, List<EmbeddedEntity>> embeddedEntityMapListOfEmbeddedEntities) {
+			Map<String, List<EmbeddedEntity>> embeddedEntityMapListOfEmbeddedEntities,
+			Map<TestDatastoreItem.Color, String> mapEnumKeys) {
 		this.intField = intField;
 		this.embeddedEntityField = embeddedEntityField;
 		this.listOfEmbeddedEntities = listOfEmbeddedEntities;
@@ -59,6 +62,7 @@ public class TestItemWithEmbeddedEntity {
 		this.embeddedMapListOfValues = embeddedMapListOfValues;
 		this.embeddedEntityMapEmbeddedEntity = embeddedEntityMapEmbeddedEntity;
 		this.embeddedEntityMapListOfEmbeddedEntities = embeddedEntityMapListOfEmbeddedEntities;
+		this.mapEnumKeys = mapEnumKeys;
 	}
 
 	@Override

--- a/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java
+++ b/spring-cloud-gcp-data-datastore/src/test/java/org/springframework/cloud/gcp/data/datastore/it/DatastoreIntegrationTests.java
@@ -24,6 +24,7 @@ import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
 import com.google.cloud.datastore.Blob;
@@ -129,6 +130,7 @@ public class DatastoreIntegrationTests extends AbstractDatastoreIntegrationTests
 		this.datastoreTemplate.deleteAll(SubEntity.class);
 		this.datastoreTemplate.deleteAll(Pet.class);
 		this.datastoreTemplate.deleteAll(PetOwner.class);
+		this.datastoreTemplate.deleteAll(Event.class);
 		this.testEntityRepository.deleteAll();
 		if (this.keyForMap != null) {
 			this.datastore.delete(this.keyForMap);
@@ -651,6 +653,26 @@ public class DatastoreIntegrationTests extends AbstractDatastoreIntegrationTests
 		assertThat(pugCount).isEqualTo(1);
 		assertThat(dogCount).isEqualTo(1);
 	}
+
+	@Test
+	public void enumKeys() {
+		Map<CommunicationChannels, String> phone = new HashMap<>();
+		phone.put(CommunicationChannels.SMS, "123456");
+
+		Map<CommunicationChannels, String> email = new HashMap<>();
+		phone.put(CommunicationChannels.EMAIL, "a@b.c");
+
+		Event event1 = new Event("event1", phone);
+		Event event2 = new Event("event2", email);
+
+		this.datastoreTemplate.saveAll(Arrays.asList(event1, event2));
+
+		waitUntilTrue(() -> this.datastoreTemplate.count(Event.class) == 2);
+
+		Collection<Event> events = this.datastoreTemplate.findAll(Event.class);
+
+		assertThat(events).containsExactlyInAnyOrder(event1, event2);
+	}
 }
 
 /**
@@ -771,4 +793,39 @@ interface DogRepository extends DatastoreRepository<Dog, Long> {
 
 	@Query("select * from Pet")
 	List<Dog> findByCustomQuery();
+}
+
+@Entity
+class Event {
+	@Id
+	private String eventName;
+
+	private Map<CommunicationChannels, String> preferences;
+
+	Event(String eventName, Map<CommunicationChannels, String> preferences) {
+		this.eventName = eventName;
+		this.preferences = preferences;
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		Event event = (Event) o;
+		return Objects.equals(this.eventName, event.eventName) &&
+				Objects.equals(this.preferences, event.preferences);
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(this.eventName, this.preferences);
+	}
+}
+
+enum CommunicationChannels {
+	EMAIL, SMS;
 }

